text/0031-integ-testing.md

```diff
@@ -0,0 +1,542 @@
+  request: AwsAssertionCall.Sqs.receiveMessage({ queue: stack.publishNotifs }),
+  returns: { Messages: [...] }
+});
+
+const testResult = repoTest.run();
+if (testResult.failures > 0) {
+  // notify the test runner to fail
+}
+```
+
+### Assertions
+
+`apptest` comes canned with a set of commonly used high level assertions.
+
+* `AwsAssertion`: execute AWS service APIs and assert the response.
+* `InvokeLambda`: invoke an AWS Lambda Function and assert its response.
+* `HttpRequest`: Execute an HTTP request against an API endpoint and assert the response.
+  Used typically to test API Gateway resources.
+
+By default, these assertions will check that the underlying call succeeds. That behavior can be extended
+to also include verification of the response.
+
+```ts
+new AwsAssertion(test, 'GetObject', {
+  request: AwsAssertionCall.S3.getObject({ ... }),
+  returns: { Data: ... },
+});
+
+new InvokeLambda(test, 'Invoke', {
+  lambda: functionArn,
+  throws: ...
+});
+```
+
+The `AwsAssertion` also allows multiple calls to be chained. The following calls Amazon S3 `GetObject` API
+based on the value returned from an Amazon DynamoDB `GetItem` call.
+
+```ts
+const response = new AwsAssertion(test, 'GetItem', {
+  request: AwsAssertionCall.DynamoDB.getItem({ ... }),
+});
+
+new AwsAssertion(test, 'GetObject', {
+  request: AwsAssertionCall.S3.getObject({ bucket: ..., key: response.Item.Payload }),
+});
+```
+
+When canned assertions are insufficient, `apptest` also provides a mechanism to write custom assertions.
+Custom assertions are simply AWS Lambda Functions, and in the CDK, these are any construct that implements `IFunction`.
+
+```ts
+import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
+
+// Checks that a standard header is present in the given S3 object
+class StdHeaderAssertion extends Function {
+  constructor(scope: Construct, id: string, props: StdHeaderAssertionProps) {
+    super(scope, id, {
+      code: Code.fromAsset('./standard-header-assertion.js'),
+      runtime: Runtime.NODEJS_12_X,
+      env: {
+        BUCKET_NAME: props.bucket.bucketName,
+        KEY: props.bucket.key
+      },
+    });
+  }
+}
+```
+
+The assertion passes or fails based on the success or failure of the Lambda's handler code.
+In both cases, it can also return an optional message, that will be made available in the test result.
+
+To invoke this assertion, you can simply use the `invokeLambda()` API -
+
+```ts
+const hdrAssertion = new StdHeaderAssertion(...);
+
+new InvokeLambda(test, {
+  lambda: hdrAssertion,
+});
+```
+
+> RFC Note: See [Appendix A](#appendix-a---assertions) for details on the underlying design.
+
+### Test snapshots
+
+On first run of a apptest test, a snapshot will be produced that *must be checked into source tree*.
+This is a snapshot of the [cloud assembly] containing both the CDK app (being tested) and the test
+assertions. The snapshot will be placed at a folder relative to the test file.
```

Comment:
I am wondering if it makes sense to package the snapshot as a single [human readable text] file instead of a directory. WDYT?